File: contracts/rust/testcore/src/testcore.rs
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use wasmlib::*;

use crate::*;
use crate::contract::*;

const CONTRACT_NAME_DEPLOYED: &str = "exampleDeployTR";
const MSG_CORE_ONLY_PANIC: &str = "========== core only =========";
const MSG_FULL_PANIC: &str = "========== panic FULL ENTRY POINT =========";
const MSG_VIEW_PANIC: &str = "========== panic VIEW =========";

pub fn func_call_on_chain(ctx: &ScFuncContext, f: &CallOnChainContext) {
    let param_int = f.params.int_value().value();

    let mut hname_contract = ctx.contract();
    if f.params.hname_contract().exists() {
        hname_contract = f.params.hname_contract().value();
    }

    let mut hname_ep = HFUNC_CALL_ON_CHAIN;
    if f.params.hname_ep().exists() {
        hname_ep = f.params.hname_ep().value();
    }

    let counter = f.state.counter();
    ctx.log(&format!("call depth = {}, hnameContract = {}, hnameEP = {}, counter = {}",
                     &f.params.int_value().to_string(),
                     &hname_contract.to_string(),
                     &hname_ep.to_string(),
                     &counter.to_string()));

    counter.set_value(counter.value() + 1);

    let parms = ScMutableMap::new();
    parms.get_int64(PARAM_INT_VALUE).set_value(param_int);
    let ret = ctx.call(hname_contract, hname_ep, Some(parms), None);
    let ret_val = ret.get_int64(RESULT_INT_VALUE);
    f.results.int_value().set_value(ret_val.value());
}

pub fn func_check_context_from_full_ep(ctx: &ScFuncContext, f: &CheckContextFromFullEPContext) {
    ctx.require(f.params.agent_id().value() == ctx.account_id(), "fail: agentID");
    ctx.require(f.params.caller().value() == ctx.caller(), "fail: caller");
    ctx.require(f.params.chain_id().value() == ctx.chain_id(), "fail: chainID");
    ctx.require(f.params.chain_owner_id().value() == ctx.chain_owner_id(), "fail: chainOwnerID");
    ctx.require(f.params.contract_creator().value() == ctx.contract_creator(), "fail: contractCreator");
}

pub fn func_do_nothing(ctx: &ScFuncContext, _f: &DoNothingContext) {
    ctx.log("doing nothing...");
}

pub fn func_get_minted_supply(ctx: &ScFuncContext, f: &GetMintedSupplyContext) {
    let minted = ctx.minted();
    let minted_colors = minted.colors();
    ctx.require(minted_colors.length() == 1, "test only supports one minted color");
    let color = minted_colors.get_color(0).value();
    let amount = minted.balance(&color);
    f.results.minted_supply().set_value(amount);
    f.results.minted_color().set_value(&color);
}

pub fn func_inc_counter(_ctx: &ScFuncContext, f: &IncCounterContext) {
    let counter = f.state.counter();
    counter.set_value(counter.value() + 1);
}

pub fn func_init(ctx: &ScFuncContext, f: &InitContext) {
    if f.params.fail().exists() {
        ctx.panic("failing on purpose");
    }
}

pub fn func_pass_types_full(ctx: &ScFuncContext, f: &PassTypesFullContext) {
    let hash = ctx.utility().hash_blake2b(PARAM_HASH.as_bytes());
    ctx.require(f.params.hash().value() == hash, "Hash wrong");
    ctx.require(f.params.int64().value() == 42, "int64 wrong");
    ctx.require(f.params.int64_zero().value() == 0, "int64-0 wrong");
    ctx.require(f.params.string().value() == PARAM_STRING, "string wrong");
    ctx.require(f.params.string_zero().value() == "", "string-0 wrong");
    ctx.require(f.params.hname().value() == ScHname::new(PARAM_HNAME), "Hname wrong");
    ctx.require(f.params.hname_zero().value() == ScHname(0), "Hname-0 wrong");
}

pub fn func_run_recursion(ctx: &ScFuncContext, f: &RunRecursionContext) {
    let depth = f.params.int_value().value();
    if depth <= 0 {
        return;
    }

    let call_on_chain = ScFuncs::call_on_chain(ctx);
    call_on_chain.params.int_value().set_value(depth - 1);
    call_on_chain.params.hname_ep().set_value(HFUNC_RUN_RECURSION);
    call_on_chain.func.call();
    let ret_val = call_on_chain.results.int_value().value();
    f.results.int_value().set_value(ret_val);
}

pub fn func_send_to_address(ctx: &ScFuncContext, f: &SendToAddressContext) {
    let balances = ScTransfers::new_transfers_from_balances(ctx.balances());
    ctx.transfer_to_address(&f.params.address().value(), balances);
}

pub fn func_set_int(_ctx: &ScFuncContext, f: &SetIntContext) {
    f.state.ints().get_int64(&f.params.name().value()).set_value(f.params.int_value().value());
}

pub fn func_test_call_panic_full_ep(ctx: &ScFuncContext, _f: &TestCallPanicFullEPContext) {
    ScFuncs::test_panic_full_ep(ctx).func.call();
}

pub fn func_test_call_panic_view_ep_from_full(ctx: &ScFuncContext, _f: &TestCallPanicViewEPFromFullContext) {
    ScFuncs::test_panic_view_ep(ctx).func.call();
}

pub fn func_test_chain_owner_id_full(ctx: &ScFuncContext, f: &TestChainOwnerIDFullContext) {
    f.results.chain_owner_id().set_value(&ctx.chain_owner_id());
}

pub fn func_test_event_log_deploy(ctx: &ScFuncContext, _f: &TestEventLogDeployContext) {
    // deploy the same contract with another name
    let program_hash = ctx.utility().hash_blake2b("testcore".as_bytes());
    ctx.deploy(&program_hash, CONTRACT_NAME_DEPLOYED, "test contract deploy log", None);
}

pub fn func_test_event_log_event_data(ctx: &ScFuncContext, _f: &TestEventLogEventDataContext) {
    ctx.event("[Event] - Testing Event...");
}

pub fn func_test_event_log_generic_data(ctx: &ScFuncContext, f: &TestEventLogGenericDataContext) {
    let event = "[GenericData] Counter Number: ".to_string() + &f.params.counter().to_string();
    ctx.event(&event);
}

pub fn func_test_panic_full_ep(ctx: &ScFuncContext, _f: &TestPanicFullEPContext) {
    ctx.panic(MSG_FULL_PANIC);
}

pub fn func_withdraw_to_chain(ctx: &ScFuncContext, f: &WithdrawToChainContext) {
    let xx = corecontracts::coreaccounts::ScFuncs::withdraw(ctx);
    xx.func.transfer_iotas(1).post_to_chain(f.params.chain_id().value());
}

pub fn view_check_context_from_view_ep(ctx: &ScViewContext, f: &CheckContextFromViewEPContext) {
    ctx.require(f.params.agent_id().value() == ctx.account_id(), "fail: agentID");
    ctx.require(f.params.chain_id().value() == ctx.chain_id(), "fail: chainID");
    ctx.require(f.params.chain_owner_id().value() == ctx.chain_owner_id(), "fail: chainOwnerID");
    ctx.require(f.params.contract_creator().value() == ctx.contract_creator(), "fail: contractCreator");
}

pub fn view_fibonacci(ctx: &ScViewContext, f: &FibonacciContext) {
    let n = f.params.int_value().value();
    if n == 0 || n == 1 {
        f.results.int_value().set_value(n);
        return;
    }

    let fib = ScFuncs::fibonacci(ctx);
    fib.params.int_value().set_value(n - 1);
    fib.func.call();
    let n1 = fib.results.int_value().value();

    fib.params.int_value().set_value(n - 2);
    fib.func.call();
    let n2 = fib.results.int_value().value();

    f.results.int_value().set_value(n1 + n2);
}

pub fn view_get_counter(_ctx: &ScViewContext, f: &GetCounterContext) {
    f.results.counter().set_value(f.state.counter().value());
}

pub fn view_get_int(ctx: &ScViewContext, f: &GetIntContext) {
    let name = f.params.name().value();
    let value = f.state.ints().get_int64(&name);
    ctx.require(value.exists(), "param 'value' not found");
    f.results.values().get_int64(&name).set_value(value.value());
}

pub fn view_just_view(ctx: &ScViewContext, _f: &JustViewContext) {
    ctx.log("doing nothing...");
}

pub fn view_pass_types_view(ctx: &ScViewContext, f: &PassTypesViewContext) {
    let hash = ctx.utility().hash_blake2b(PARAM_HASH.as_bytes());
    ctx.require(f.params.hash().value() == hash, "Hash wrong");
    ctx.require(f.params.int64().value() == 42, "int64 wrong");
    ctx.require(f.params.int64_zero().value() == 0, "int64-0 wrong");
    ctx.require(f.params.string().value() == PARAM_STRING, "string wrong");
    ctx.require(f.params.string_zero().value() == "", "string-0 wrong");
    ctx.require(f.params.hname().value() == ScHname::new(PARAM_HNAME), "Hname wrong");
    ctx.require(f.params.hname_zero().value() == ScHname(0), "Hname-0 wrong");
}

pub fn view_test_call_panic_view_ep_from_view(ctx: &ScViewContext, _f: &TestCallPanicViewEPFromViewContext) {
    ScFuncs::test_panic_view_ep(ctx).func.call();
}

pub fn view_test_chain_owner_id_view(ctx: &ScViewContext, f: &TestChainOwnerIDViewContext) {
    f.results.chain_owner_id().set_value(&ctx.chain_owner_id());
}

pub fn view_test_panic_view_ep(ctx: &ScViewContext, _f: &TestPanicViewEPContext) {
    ctx.panic(MSG_VIEW_PANIC);
}

pub fn view_test_sandbox_call(ctx: &ScViewContext, f: &TestSandboxCallContext) {
    let get_chain_info = corecontracts::coregovernance::ScFuncs::get_chain_info(ctx);
    get_chain_info.func.call();
    f.results.sandbox_call().set_value(&get_chain_info.results.description().value());
}

pub fn func_test_block_context1(ctx: &ScFuncContext, _f: &TestBlockContext1Context) {
    ctx.panic(MSG_CORE_ONLY_PANIC);
}

pub fn func_test_block_context2(ctx: &ScFuncContext, _f: &TestBlockContext2Context) {
    ctx.panic(MSG_CORE_ONLY_PANIC);
}

pub fn view_get_string_value(ctx: &ScViewContext, _f: &GetStringValueContext) {
    ctx.panic(MSG_CORE_ONLY_PANIC);
}

pub fn func_spawn(ctx: &ScFuncContext, f: &SpawnContext) {
    let spawn_name = SC_NAME.to_string() + "_spawned";
    let spawn_descr = "spawned contract description";
    ctx.deploy(&f.params.prog_hash().value(), &spawn_name, spawn_descr, None);

    let spawn_hname = ScHname::new(&spawn_name);
    for _i in 0..5 {
        ctx.call(spawn_hname, HFUNC_INC_COUNTER, None, None);
    }
}

File: contracts/wasm/testcore/ts/testcore/testcore.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import * as wasmlib from "wasmlib"
import * as wasmtypes from "wasmlib/wasmtypes";
import * as coreaccounts from "wasmlib/coreaccounts"
import * as coregovernance from "wasmlib/coregovernance"
import * as sc from "./index";

const CONTRACT_NAME_DEPLOYED = "exampleDeployTR";
const MSG_CORE_ONLY_PANIC = "========== core only =========";
const MSG_FULL_PANIC = "========== panic FULL ENTRY POINT =========";
const MSG_VIEW_PANIC = "========== panic VIEW =========";

export function funcCallOnChain(ctx: wasmlib.ScFuncContext, f: sc.CallOnChainContext): void {
    let paramInt = f.params.intValue().value();

    let hnameContract = ctx.contract();
    if (f.params.hnameContract().exists()) {
        hnameContract = f.params.hnameContract().value();
    }

    let hnameEP = sc.HFuncCallOnChain;
    if (f.params.hnameEP().exists()) {
        hnameEP = f.params.hnameEP().value();
    }

    let counter = f.state.counter();
    ctx.log("call depth = " + f.params.intValue().toString() +
        ", hnameContract = " + hnameContract.toString() +
        ", hnameEP = " + hnameEP.toString() +
        ", counter = " + counter.toString())

    counter.setValue(counter.value() + 1);

    let params = new wasmlib.ScDict([]);
    const key = wasmtypes.stringToBytes(sc.ParamIntValue);
    params.set(key, wasmtypes.int64ToBytes(paramInt))
    let ret = ctx.call(hnameContract, hnameEP, params, null);
    let retVal = wasmtypes.int64FromBytes(ret.get(key));
    f.results.intValue().setValue(retVal);
}

export function funcCheckContextFromFullEP(ctx: wasmlib.ScFuncContext, f: sc.CheckContextFromFullEPContext): void {
    ctx.require(f.params.agentID().value().equals(ctx.accountID()), "fail: agentID");
    ctx.require(f.params.caller().value().equals(ctx.caller()), "fail: caller");
    ctx.require(f.params.chainID().value().equals(ctx.chainID()), "fail: chainID");
    ctx.require(f.params.chainOwnerID().value().equals(ctx.chainOwnerID()), "fail: chainOwnerID");
    ctx.require(f.params.contractCreator().value().equals(ctx.contractCreator()), "fail: contractCreator");
}

export function funcDoNothing(ctx: wasmlib.ScFuncContext, f: sc.DoNothingContext): void {
    ctx.log("doing nothing...");
}

// export function funcGetMintedSupply(ctx: wasmlib.ScFuncContext, f: sc.GetMintedSupplyContext): void {
//     let minted = ctx.minted();
//     let mintedColors = minted.colors();
//     ctx.require(mintedColors.length == 1, "test only supports one minted color");
//     let color = mintedColors[0];
//     let amount = minted.balance(color);
//     f.results.mintedSupply().setValue(amount);
//     f.results.mintedColor().setValue(color);
// }

export function funcIncCounter(ctx: wasmlib.ScFuncContext, f: sc.IncCounterContext): void {
    let counter = f.state.counter();
    counter.setValue(counter.value() + 1);
}

export function funcInit(ctx: wasmlib.ScFuncContext, f: sc.InitContext): void {
    if (f.params.fail().exists()) {
        ctx.panic("failing on purpose");
    }
}

export function funcPassTypesFull(ctx: wasmlib.ScFuncContext, f: sc.PassTypesFullContext): void {
    let hash = ctx.utility().hashBlake2b(wasmtypes.stringToBytes(sc.ParamHash));
    ctx.require(f.params.hash().value().equals(hash), "Hash wrong");
    ctx.require(f.params.int64().value() == 42, "int64 wrong");
    ctx.require(f.params.int64Zero().value() == 0, "int64-0 wrong");
    ctx.require(f.params.string().value() == sc.ParamString, "string wrong");
    ctx.require(f.params.stringZero().value() == "", "string-0 wrong");
    ctx.require(f.params.hname().value().equals(ctx.utility().hname(sc.ParamHname)), "Hname wrong");
    ctx.require(f.params.hnameZero().value().equals(new wasmtypes.ScHname(0)), "Hname-0 wrong");
}

export function funcRunRecursion(ctx: wasmlib.ScFuncContext, f: sc.RunRecursionContext): void {
    let depth = f.params.intValue().value();
    if (depth <= 0) {
        return;
    }

    let callOnChain = sc.ScFuncs.callOnChain(ctx);
    callOnChain.params.intValue().setValue(depth - 1);
    callOnChain.params.hnameEP().setValue(sc.HFuncRunRecursion);
    callOnChain.func.call();
    let retVal = callOnChain.results.intValue().value();
    f.results.intValue().setValue(retVal);
}

export function funcSendToAddress(ctx: wasmlib.ScFuncContext, f: sc.SendToAddressContext): void {
    // let transfer = wasmlib.ScTransfers.fromBalances(ctx.balances());
    // ctx.send(f.params.address().value(), transfer);
}

export function funcSetInt(ctx: wasmlib.ScFuncContext, f: sc.SetIntContext): void {
    f.state.ints().getInt64(f.params.name().value()).setValue(f.params.intValue().value());
}

export function funcSpawn(ctx: wasmlib.ScFuncContext, f: sc.SpawnContext): void {
    let programHash = f.params.progHash().value();
    let spawnName = sc.ScName + "_spawned";
    let spawnDescr = "spawned contract description";
    ctx.deployContract(programHash, spawnName, spawnDescr, null);

    let spawnHname = ctx.utility().hname(spawnName);
    for (let i = 0; i < 5; i++) {
        ctx.call(spawnHname, sc.HFuncIncCounter, null, null);
    }
}

export function funcTestBlockContext1(ctx: wasmlib.ScFuncContext, f: sc.TestBlockContext1Context): void {
    ctx.panic(MSG_CORE_ONLY_PANIC);
}

export function funcTestBlockContext2(ctx: wasmlib.ScFuncContext, f: sc.TestBlockContext2Context): void {
    ctx.panic(MSG_CORE_ONLY_PANIC);
}

export function funcTestCallPanicFullEP(ctx: wasmlib.ScFuncContext, f: sc.TestCallPanicFullEPContext): void {
    sc.ScFuncs.testPanicFullEP(ctx).func.call();
}

export function funcTestCallPanicViewEPFromFull(ctx: wasmlib.ScFuncContext, f: sc.TestCallPanicViewEPFromFullContext): void {
    sc.ScFuncs.testPanicViewEP(ctx).func.call();
}

export function funcTestChainOwnerIDFull(ctx: wasmlib.ScFuncContext, f: sc.TestChainOwnerIDFullContext): void {
    f.results.chainOwnerID().setValue(ctx.chainOwnerID());
}

export function funcTestEventLogDeploy(ctx: wasmlib.ScFuncContext, f: sc.TestEventLogDeployContext): void {
    // deploy the same contract with another name
    let programHash = ctx.utility().hashBlake2b(wasmtypes.stringToBytes("testcore"));
    ctx.deployContract(programHash, CONTRACT_NAME_DEPLOYED, "test contract deploy log", null);
}

export function funcTestEventLogEventData(ctx: wasmlib.ScFuncContext, f: sc.TestEventLogEventDataContext): void {
    ctx.event("[Event] - Testing Event...");
}

export function funcTestEventLogGenericData(ctx: wasmlib.ScFuncContext, f: sc.TestEventLogGenericDataContext): void {
    let event = "[GenericData] Counter Number: ".toString() + f.params.counter().toString();
    ctx.event(event);
}

export function funcTestPanicFullEP(ctx: wasmlib.ScFuncContext, f: sc.TestPanicFullEPContext): void {
    ctx.panic(MSG_FULL_PANIC);
}

// export function funcWithdrawToChain(ctx: wasmlib.ScFuncContext, f: sc.WithdrawToChainContext): void {
//     let xx = coreaccounts.ScFuncs.withdraw(ctx);
//     xx.func.postToChain(f.params.chainID().value());
// }

export function viewCheckContextFromViewEP(ctx: wasmlib.ScViewContext, f: sc.CheckContextFromViewEPContext): void {
    ctx.require(f.params.agentID().value().equals(ctx.accountID()), "fail: agentID");
    ctx.require(f.params.chainID().value().equals(ctx.chainID()), "fail: chainID");
    ctx.require(f.params.chainOwnerID().value().equals(ctx.chainOwnerID()), "fail: chainOwnerID");
    ctx.require(f.params.contractCreator().value().equals(ctx.contractCreator()), "fail: contractCreator");
}

export function viewFibonacci(ctx: wasmlib.ScViewContext, f: sc.FibonacciContext): void {
    let n = f.params.intValue().value();
    if (n == 0 || n == 1) {
        f.results.intValue().setValue(n);
        return;
    }

    let fib = sc.ScFuncs.fibonacci(ctx);
    fib.params.intValue().setValue(n - 1);
    fib.func.call();
    let n1 = fib.results.intValue().value();

    fib.params.intValue().setValue(n - 2);
    fib.func.call();
    let n2 = fib.results.intValue().value();

    f.results.intValue().setValue(n1 + n2);
}

export function viewGetCounter(ctx: wasmlib.ScViewContext, f: sc.GetCounterContext): void {
    f.results.counter().setValue(f.state.counter().value());
}

export function viewGetInt(ctx: wasmlib.ScViewContext, f: sc.GetIntContext): void {
    let name = f.params.name().value();
    let value = f.state.ints().getInt64(name);
    ctx.require(value.exists(), "param '" + name + "' not found");
    f.results.values().getInt64(name).setValue(value.value());
}

export function viewGetStringValue(ctx: wasmlib.ScViewContext, f: sc.GetStringValueContext): void {
    ctx.panic(MSG_CORE_ONLY_PANIC);
}

export function viewJustView(ctx: wasmlib.ScViewContext, f: sc.JustViewContext): void {
    ctx.log("doing nothing...");
}

export function viewPassTypesView(ctx: wasmlib.ScViewContext, f: sc.PassTypesViewContext): void {
    let hash = ctx.utility().hashBlake2b(wasmtypes.stringToBytes(sc.ParamHash));
    ctx.require(f.params.hash().value().equals(hash), "Hash wrong");
    ctx.require(f.params.int64().value() == 42, "int64 wrong");
    ctx.require(f.params.int64Zero().value() == 0, "int64-0 wrong");
    ctx.require(f.params.string().value() == sc.ParamString, "string wrong");
    ctx.require(f.params.stringZero().value() == "", "string-0 wrong");
    ctx.require(f.params.hname().value().equals(ctx.utility().hname(sc.ParamHname)), "Hname wrong");
    ctx.require(f.params.hnameZero().value().equals(new wasmtypes.ScHname(0)), "Hname-0 wrong");
}

export function viewTestCallPanicViewEPFromView(ctx: wasmlib.ScViewContext, f: sc.TestCallPanicViewEPFromViewContext): void {
    sc.ScFuncs.testPanicViewEP(ctx).func.call();
}

export function viewTestChainOwnerIDView(ctx: wasmlib.ScViewContext, f: sc.TestChainOwnerIDViewContext): void {
    f.results.chainOwnerID().setValue(ctx.chainOwnerID());
}

export function viewTestPanicViewEP(ctx: wasmlib.ScViewContext, f: sc.TestPanicViewEPContext): void {
    ctx.panic(MSG_VIEW_PANIC);
}

export function viewTestSandboxCall(ctx: wasmlib.ScViewContext, f: sc.TestSandboxCallContext): void {
    let getChainInfo = coregovernance.ScFuncs.getChainInfo(ctx);
    getChainInfo.func.call();
    f.results.sandboxCall().setValue(getChainInfo.results.description().value());
}

export function funcClaimAllowance(ctx: wasmlib.ScFuncContext, f: sc.ClaimAllowanceContext): void {
}

export function funcEstimateMinDust(ctx: wasmlib.ScFuncContext, f: sc.EstimateMinDustContext): void {
}

export function funcInfiniteLoop(ctx: wasmlib.ScFuncContext, f: sc.InfiniteLoopContext): void {
    for (; ;) {
        // do nothing, just waste gas
    }
}

export function funcPingAllowanceBack(ctx: wasmlib.ScFuncContext, f: sc.PingAllowanceBackContext): void {
}

export function funcSendLargeRequest(ctx: wasmlib.ScFuncContext, f: sc.SendLargeRequestContext): void {
}

export function funcSendNFTsBack(ctx: wasmlib.ScFuncContext, f: sc.SendNFTsBackContext): void {
}

export function funcSplitFunds(ctx: wasmlib.ScFuncContext, f: sc.SplitFundsContext): void {
    let iotas = ctx.allowance().balance(wasmtypes.IOTA);
    const address = ctx.caller().address();
    const transfer = wasmlib.ScTransfers.iotas(200);
    for (; iotas >= 200; iotas -= 200) {
        ctx.transferAllowed(ctx.accountID(), transfer, false);
        ctx.send(address, transfer);
    }
}

export function funcSplitFundsNativeTokens(ctx: wasmlib.ScFuncContext, f: sc.SplitFundsNativeTokensContext): void {
    let iotas = ctx.allowance().balance(wasmtypes.IOTA);
    const address = ctx.caller().address();
    let transfer = wasmlib.ScTransfers.iotas(iotas);
    ctx.transferAllowed(ctx.accountID(), transfer, false);
    const colors = ctx.allowance().colors();
    for (let i = 0; i < colors.length; i++) {
        const token = colors[i];
        if (token.equals(wasmtypes.IOTA)) {
            continue;
        }
        transfer = wasmlib.ScTransfers.transfer(token, 1);
        let tokens = ctx.allowance().balance(token);
        for (; tokens >= 1; tokens--) {
            ctx.transferAllowed(ctx.accountID(), transfer, false);
            ctx.send(address, transfer);
        }
    }
}

export function funcWithdrawFromChain(ctx: wasmlib.ScFuncContext, f: sc.WithdrawFromChainContext): void {
}

export function viewInfiniteLoopView(ctx: wasmlib.ScViewContext, f: sc.InfiniteLoopViewContext): void {
    for (; ;) {
        // do nothing, just waste gas
    }
}
